File: src/c++/exercises/02_Images/ConvertToGray/ConvertToGray.cpp
/*****************************************************************************************************
 * Lecture sample code.
 * Convert color to gray image.
 *****************************************************************************************************
 * Author: Marc Hensel, http://www.haw-hamburg.de/marc-hensel
 * Project: https://github.com/MarcOnTheMoon/imaging_learners/
 * Copyright: 2024, Marc Hensel
 * Version: 2024.10.09
 * License: CC BY-NC-SA 4.0, see https://creativecommons.org/licenses/by-nc-sa/4.0/deed.en
 *****************************************************************************************************/

/* Compiler settings */
#define _CRT_SECURE_NO_WARNINGS		// Enable getenv()

/* Include files */
#include <iostream>
#include <string>
#include <opencv2/opencv.hpp>

/* Defines */
#define DATA_ROOT_PATH getenv("ImagingData")  // Read environment variable
#define INPUT_IMAGE "/images/misc/Docks.jpg"

/* Namespaces */
using namespace cv;
using namespace std;

/* Main function */
int main()
{
	// Load image from file
	string inputImagePath = string(DATA_ROOT_PATH).append(INPUT_IMAGE);
	Mat colorImage = imread(inputImagePath);
	Mat grayImage;

	if (colorImage.empty()) {
		cout << "[ERROR] Cannot open image: " << inputImagePath << endl;
		return 0;
	}

	// Display image in named window and wait for keypress
	imshow("Image", colorImage);
	waitKey(0);

	// Convert image to grayscale and display in window
	cvtColor(colorImage, grayImage, COLOR_BGR2GRAY);
	imshow("Image", grayImage);

	// Wait for keypress and terminate
	waitKey(0);
	return 0;
}

File: src/python/exercises/02_Images/ConvertToGray.py
"""
Convert color to gray image.

@author: Marc Hensel
@contact: http://www.haw-hamburg.de/marc-hensel

@copyright: 2025, Marc Hensel
@version: 2025.03.30
@license: CC BY-NC-SA 4.0, see https://creativecommons.org/licenses/by-nc-sa/4.0/deed.en
"""

import cv2
import os

# Configuration
image_path = os.environ['ImagingData'] + '/images/misc/Docks.jpg'

# Load image from file
color_image = cv2.imread(image_path)

# Display image in named window and wait for keypress
cv2.imshow('Image', color_image)
cv2.waitKey(0)

# Convert image to grayscale and display in window
gray_image = cv2.cvtColor(color_image, cv2.COLOR_BGR2GRAY)
cv2.imshow('Image', gray_image)

# Wait for keypress
cv2.waitKey(0)

# Close all named windows
cv2.destroyAllWindows()